### PR TITLE
fix #4080

### LINKED ISSUE #4080: current master breaks network devices in udev on 32 bit
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

Until a few days ago our integration tests have worked fine in master. But now i386 consistently fails `tests/networkd-test.py` on i386 (for example, in https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-xenial-pitti-systemd-semaphore/xenial/i386/s/systemd-upstream/20160901_125756@/log.gz)

This probably started with PR #4017 which got merged with the regression on i386. @ssahani , do you have time to check this on 32 bit?

### PR DESCRIPTION
**Pull request by ssahani**
ethtool_sset_info moving the allocation to heap and adding some
extra space to it.

also fix valgrind warning

```
Unloaded link configuration context.
==31690==
==31690== HEAP SUMMARY:
==31690==     in use at exit: 8,192 bytes in 2 blocks
==31690==   total heap usage: 431 allocs, 429 frees, 321,164 bytes allocated
==31690==
==31690== 4,096 bytes in 1 blocks are still reachable in loss record 1 of 2
==31690==    at 0x4C2BBAD: malloc (vg_replace_malloc.c:299)
==31690==    by 0x166B32: mempool_alloc_tile (mempool.c:62)
==31690==    by 0x166BBC: mempool_alloc0_tile (mempool.c:81)
==31690==    by 0x15B8FC: hashmap_base_new (hashmap.c:732)
==31690==    by 0x15B9F7: internal_hashmap_new (hashmap.c:766)
==31690==    by 0x151291: conf_files_list_strv_internal (conf-files.c:103)
==31690==    by 0x1514BA: conf_files_list_strv (conf-files.c:135)
==31690==    by 0x13A1CF: link_config_load (link-config.c:227)
==31690==    by 0x135B68: builtin_net_setup_link_init
(udev-builtin-net_setup_link.c:77)
==31690==    by 0x1306B3: udev_builtin_init (udev-builtin.c:57)
==31690==    by 0x11E984: adm_builtin (udevadm-test-builtin.c:72)
==31690==    by 0x117B4D: run_command (udevadm.c:75)

```

Fixes #4080

### CODE REVIEW

File: src/udev/net/ethtool-util.c [hunk 1]
```diff
@@ -217,24 +217,27 @@ int ethtool_set_wol(int *fd, const char *ifname, WakeOnLan wol) {
...
         _cleanup_free_ struct ethtool_gstrings *strings = NULL;
-        struct ethtool_sset_info info =  {
-                .cmd = ETHTOOL_GSSET_INFO,
-                .reserved = 0,
-                .sset_mask = 1ULL << stringset_id,
-        };
+        struct {
+                struct ethtool_sset_info info;
+                uint32_t space;
+        } buffer;
```
Comment (poettering) [inline src/udev/net/ethtool-util.c]: please leave the initialization right here, i.e.

``` c
struct {
        struct ethtool_sset_info info;
        uint32_t space;
} buffer = {
        .info = {
                .cmd = ETHTOOL_GSSET_INFO,
                .sset_mask = UINT64_C(1) << stringset_id,
        },
};
```

(no need to initialize `.reserved` explicitly btw, the compiler automatically fills in the "gaps" with zeroes if you initialize a structure like this)

Comment (ssahani) [reply src/udev/net/ethtool-util.c]: Thanks done .

**Comment by poettering** [comment]
lgtm. Let's see if this also makes CI happy.

**Comment by martinpitt** [comment]
Yay all green again, thanks @ssahani !

**Pull request merged**